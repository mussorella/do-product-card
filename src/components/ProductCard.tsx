import styles from "../styles/styles.module.css";

import React, {  createContext } from "react";
import { useProduct } from "../hooks/useProduct";
import { Product, ProductContextProps, InitialValues, onChangeArgs, ProductCardHandlers } from '../interfaces/interfaces';






export const ProductContext=createContext({} as ProductContextProps);
const {Provider}=ProductContext;//el provider es el proveedor de info


export interface Props{
    product: Product;//aca obliga que el proucto sea enviado como tal
    //children?: ReactElement| ReactElement[];//es un elemento de react
    children:(args: ProductCardHandlers)=>JSX.Element;//lo q me permite hacerlo jsx y no comoa rriba es que ahora puedo agregarle funciones o patrones o componentes a esta interfaz por ejemplo u string y mandarlo a llamar a otro lado
    className?: string;
    style?: React.CSSProperties;
    onChange?:(args:onChangeArgs)=>void;
    value?:number;
    initialValues?: InitialValues
  }


export const ProductCard = ({children,product, className, style, onChange, value, initialValues}: Props) => {
 
    const {counter,increaseBy, maxCount, isMaxCountReached, reset}= useProduct({
        onChange,
         product, 
         value,
         initialValues
        })


    return(//pongo el provider aca para que exporte los datos 
 <Provider value={{
  counter,
  increaseBy,
  product,
  maxCount
 }}>

    <div className={`${styles.productCard}  ${className}`}
    style={style}>
        {children({
            count:counter,
            isMaxCountReached,
            maxCount:initialValues?.maxCount,
            product,
            increaseBy,
            reset,
        })}
        
    </div>

 </Provider>
 
)
    
}




