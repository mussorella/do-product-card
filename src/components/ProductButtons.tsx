import React,{ useCallback, useContext } from "react"
import { ProductContext } from "./ProductCard"
import styles from "../styles/styles.module.css";




export interface Props{
className?: string
style?: React.CSSProperties
}


export const ProductButtons=({className, style}: Props)=>{//desestructuro las props de arriba
    
    const {increaseBy,counter, maxCount}=useContext(ProductContext)
  

    const isMaxReached=useCallback(
      () => !!maxCount && counter ===maxCount,//si tiene un valor regresa maxcount sino rturna falso y no sigue
      [counter, maxCount],
    );
    

    return(
      <div 
      className={`${styles.buttonsContainer} ${className}`}
      style={style}
      >
              
              <button className={styles.buttonMinus} onClick={()=> increaseBy(-1)}> - </button>
              <div className={styles.countLabel}> {counter} </div>
              <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={()=> increaseBy(+1)}> + </button>
              
          </div>
    )
  }