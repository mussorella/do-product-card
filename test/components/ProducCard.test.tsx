import * as React from 'react';
import renderer from 'react-test-renderer';

import { product2 } from '../data/products';
import { ProductCard } from '../../src/components/ProductCard';

const {act}=renderer;

describe('ProductCard', () => { 

test('debe de mostrar el componente correactamente', () => { 
     


const wrapper= renderer.create(
    <ProductCard product={product2} >
{
    ()=>(
        <h1>ProductCard</h1>
    )
}
    </ProductCard>
    
)
expect(wrapper.toJSON()).toMatchSnapshot()


    });



    test('Debe de incrementar el contador', () => { 


        const wrapper= renderer.create(
            <ProductCard product={product2} >
        {
            ({count, increaseBy})=>(
                <>
                <h1>ProductCard</h1>
                <span>{count}</span>
                <button onClick={()=> increaseBy(1)}></button>
                </>
            )
            
        }
            </ProductCard>
            
        );
        let tree = wrapper.toJSON()
        expect(tree).toMatchSnapshot();


        act(()=>{
            (tree as any).children[2].props.onClick()//estoiy ejecutando el onClick
        })

       
        tree=wrapper.toJSON()
        expect((tree as any).children[1].children[0]).toBe('1')

     })

})