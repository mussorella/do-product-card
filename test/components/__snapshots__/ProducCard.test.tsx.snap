// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Debe de incrementar el contador 1`] = `
<div
  className="productCard  undefined"
>
  <h1>
    ProductCard
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`ProductCard debe de mostrar el componente correactamente 1`] = `
<div
  className="productCard  undefined"
>
  <h1>
    ProductCard
  </h1>
</div>
`;
